proof/refine/ARM/KHeap_R.thy

```diff
@@ -147,60 +269,57 @@ lemma objBitsKO_bounded[simp]:
-  apply (rule ps_clear_lookupAround2, simp+)
-   apply (simp add: p_assoc_help)
-   apply (erule is_aligned_no_wrap'[OF _ word32_power_less_1[OF objBitsKO_bounded]])
-  apply (clarsimp simp: objBits_simps' simp del: lookupAround2_same1 split: option.splits)
-  done
-
 lemma typ_at_to_obj_at':
   "typ_at' (koType (TYPE ('a :: pspace_storable))) p s
      = obj_at' (\<top> :: 'a \<Rightarrow> bool) p s"
   by (simp add: typ_at'_def obj_at'_real_def project_koType[symmetric])
 
 lemmas typ_at_to_obj_at_arches
   = typ_at_to_obj_at'[where 'a=pte, simplified]
-    typ_at_to_obj_at' [where 'a=pde, simplified]
+    typ_at_to_obj_at'[where 'a=pde, simplified]
     typ_at_to_obj_at'[where 'a=asidpool, simplified]
     typ_at_to_obj_at'[where 'a=user_data, simplified]
     typ_at_to_obj_at'[where 'a=user_data_device, simplified]
 
 lemmas page_table_at_obj_at'
   = page_table_at'_def[unfolded typ_at_to_obj_at_arches]
 
-
-lemma corres_get_tcb [corres]:
-  "corres (tcb_relation \<circ> the) (tcb_at t) (tcb_at' t) (gets (get_tcb t)) (getObject t)"
-  apply (rule corres_no_failI)
-   apply wp
-  apply (clarsimp simp add: gets_def get_def return_def bind_def get_tcb_def)
-  apply (frule in_inv_by_hoareD [OF getObject_tcb_inv])
-  apply (clarsimp simp add: obj_at_def is_tcb obj_at'_def projectKO_def
-                            projectKO_opt_tcb split_def
-                            getObject_def loadObject_default_def in_monad)
-  apply (case_tac obj)
-   apply (simp_all add: fail_def return_def)
-  apply (case_tac bb)
-   apply (simp_all add: fail_def return_def)
-  apply (clarsimp simp add: state_relation_def pspace_relation_def)
-  apply (drule bspec)
-   apply clarsimp
-   apply blast
-  apply (clarsimp simp add: other_obj_relation_def
-                            lookupAround2_known1)
-  done
+method readObject_obj_at'_method
+  =  clarsimp simp: readObject_def obind_def in_omonad split_def loadObject_default_def
+                    obj_at'_def projectKOs objBits_simps' scBits_pos_power2
+             split: option.splits if_split_asm
+
+lemma readObject_misc_ko_at':
+  "readObject p s = Some (tcb :: tcb) \<Longrightarrow> ko_at' tcb p s"
+  "readObject p s = Some (ep :: endpoint) \<Longrightarrow> ko_at' ep p s"
+  "readObject p s = Some (ntfn :: notification) \<Longrightarrow> ko_at' ntfn p s"
+  "readObject p s = Some (reply :: reply) \<Longrightarrow> ko_at' reply p s"
+  "readObject p s = Some (sc :: sched_context) \<Longrightarrow> ko_at' sc p s"
+   by readObject_obj_at'_method+
+
+lemmas readObject_ko_at'_tcb  [simp, dest!] = readObject_misc_ko_at'(1)
+lemmas readObject_ko_at'_ep   [simp, dest!] = readObject_misc_ko_at'(2)
+lemmas readObject_ko_at'_ntfn [simp, dest!] = readObject_misc_ko_at'(3)
+lemmas readObject_ko_at'_reply[simp, dest!] = readObject_misc_ko_at'(4)
+lemmas readObject_ko_at'_sc   [simp, dest!] = readObject_misc_ko_at'(5)
+
+lemma readObject_misc_obj_at':
+  "bound (readObject p s ::tcb option) \<Longrightarrow> tcb_at' p s"
```

Comment:
Would it be best to eithe add a space after `::` or remove the one before?

Reply:
I think we should standardise on space before and after, at least if it is a longer type name like this or something with a type parameter. We could have exceptions for single-letter type variables, i.e. `x::'a` still reads fine, but `x::'a word` visually suggests the wrong grouping `(x::'a) (word)`, whereas `x :: 'a word` is at least neutral.

There is also the unicode symbol that has the two colons close together. That one could be used without extra spaces, because the symbol basically has them, but I'd prefer the ASCII version of the symbol usually.